pkg/k8s/identitybackend/identity.go

```diff
@@ -325,33 +328,55 @@ func (c *crdBackend) GetByID(ctx context.Context, id idpool.ID) (allocator.Alloc
 
 	obj, exists, err := c.Store.Get(identityTemplate)
 	if err != nil {
-		return nil, err
+		return nil, exists, err
 	}
 	if !exists {
-		return nil, nil
+		return nil, exists, nil
 	}
 
 	identity, ok := obj.(*types.Identity)
 	if !ok {
-		return nil, fmt.Errorf("invalid object")
+		return nil, false, fmt.Errorf("invalid object")
+	}
+	return identity, true, nil
+}
+
+// GetByID returns the key associated with an ID. Returns nil if no key is
+// associated with the ID.
+// Note: the lock field is not supported with the k8s CRD allocator.
+func (c *crdBackend) GetByID(ctx context.Context, id idpool.ID) (allocator.AllocatorKey, error) {
+	identity, exists, err := c.getById(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, nil
 	}
 
 	return c.KeyType.PutKeyFromMap(identity.SecurityLabels), nil
 }
 
-// Release dissociates this node from using the identity bound to key. When an
-// identity has no references it may be garbage collected.
-func (c *crdBackend) Release(ctx context.Context, key allocator.AllocatorKey) (err error) {
-	identity := c.get(ctx, key)
-	if identity == nil {
+// Release dissociates this node from using the identity bound to the given ID.
+// When an identity has no references it may be garbage collected.
+func (c *crdBackend) Release(ctx context.Context, id idpool.ID, key allocator.AllocatorKey) (err error) {
+	identity, exists, err := c.getById(ctx, id)
+	if err != nil {
+		return err
+	}
+	if !exists || identity == nil {
 		return fmt.Errorf("unable to release identity %s: identity does not exist", key)
 	}
 
 	if _, ok := identity.Status.Nodes[c.NodeName]; !ok {
 		return fmt.Errorf("unable to release identity %s: identity is unused", key)
 	}
 
-	delete(identity.Status.Nodes, c.NodeName)
+	if identity.Status.Nodes == nil {
```

Comment:
Shouldn't this check need to happen before line 370? `identity.Status.Nodes` is checked for `nil` after already accessing it.

Reply:
Actually this is a bug, I should have left the `delete(identityCopy.Status.Nodes, c.NodeName)`. We should not be checking at the `identity.Status.Nodes == nil` at all because the Nodes should contain at least this node's name.